kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -504,76 +530,116 @@ protected void handleTopicMetadataRequest(KafkaHeaderAndRequest metadataHar,
-                    final String fullTopicName = new KopTopic(topic).getFullName();
-
-                    // get partition numbers for each topic.
-                    // If topic doesn't exist and allowAutoTopicCreation is enabled, the topic will be created first.
-                    getPartitionedTopicMetadataAsync(fullTopicName)
-                        .whenComplete((partitionedTopicMetadata, throwable) -> {
-                            if (throwable != null) {
-                                if (throwable instanceof PulsarAdminException.NotFoundException) {
-                                    if (kafkaConfig.isAllowAutoTopicCreation()
-                                            && metadataRequest.allowAutoTopicCreation()) {
-                                        log.info("[{}] Request {}: Topic {} doesn't exist, auto create it with {} "
-                                                        + "partitions", ctx.channel(), metadataHar.getHeader(),
-                                                topic, defaultNumPartitions);
-                                        admin.topics().createPartitionedTopicAsync(fullTopicName, defaultNumPartitions)
-                                                .whenComplete((ignored, e) -> {
-                                                    if (e == null) {
-                                                        addTopicPartition.accept(topic, defaultNumPartitions);
-                                                    } else {
-                                                        log.error("[{}] Failed to create partitioned topic {}",
-                                                                ctx.channel(), topic, e);
-                                                        completeOneTopic.run();
+            final BiConsumer<String, String> completeOneAuthFailedTopic = (topic, fullTopicName) -> {
+                allTopicMetadata.add(new TopicMetadata(
+                        Errors.TOPIC_AUTHORIZATION_FAILED,
+                        topic,
+                        isInternalTopic(fullTopicName),
+                        Collections.emptyList()));
+                completeOneTopic.run();
+            };
+
+            requestTopics.forEach(topic -> {
+                final String fullTopicName = new KopTopic(topic).getFullName();
+
+                authorize(AclOperation.DESCRIBE, Resource.of(ResourceType.TOPIC, fullTopicName))
+                    .whenComplete((authorized, ex) -> {
+                        if (ex != null) {
+                            // Authentication failed
+                            completeOneAuthFailedTopic.accept(topic, fullTopicName);
+                            return;
+                        }
+                        if (!authorized) {
+                            // Permission denied
+                            completeOneAuthFailedTopic.accept(topic, fullTopicName);
+                            return;
+                        }
```

Comment:
Add logs here to print the error message.

Reply:
In `SimpleAclAuthorizer#128` already have error message, do we need add error message here ?

Reply:
Yes. And I think we should not add error message to the API that returns a future. Though some existing code doesn't obey this rule.

When we design a API, it should be like:

```java
CompletableFuture<T> getSomething(String key) {
    return someService.findAsync(key);
}
```

The caller is responsible to determine which logs should be printed:

```java
String key = "xxx";
getSomething(key).whenComplete(/* ... */); // or exceptionally
```

Reply:
A exception case is, if we want to swallow the exception and complete the future with `null` instead of the exception. We must print the exception in the API implementation.

Reply:
Ok, I will change `SimpleAclAuthorizer#128` logs level to debug, and add log when authorize failed instead.